### PR TITLE
Implement LION optimizer

### PR DESCRIPTION
**Pull request by Alexander-Ploskin**
## 1. What Did We Implement?

We introduced the Lion optimizer [(Chen et al., 2023)](https://arxiv.org/abs/2302.06675), a momentum-based method that updates parameters using the sign of the momentum rather than scaling by the variance of recent gradients (as in Adam/AdamW). Concretely, each parameter $\theta$ is updated as follows:

---

### Lion

Given: $\beta_1, \beta_2, \lambda, \eta$  
Initialize: $\theta_0, m_0 \leftarrow 0$  
while not converged:

1. $g_t \leftarrow \nabla_\theta f(\theta_{t-1})$

2. Update model parameters
   - Compute a temporary momentum $c_t$:

     $$c_t \leftarrow \beta_1 m_{t-1} + (1 - \beta_1) g_t$$
   - Parameter update (sign-based + weight decay):
   
     $$\theta_t \leftarrow \theta_{t-1} - \eta \bigl(\mathrm{sign}(c_t) + \lambda \,\theta_{t-1}\bigr)$$

3. Update EMA of gradients (for logging or optional use):

   $$m_t \leftarrow \beta_2 m_{t-1} + (1 - \beta_2) g_t$$

---

### AdamW

Given: $\beta_1, \beta_2, \lambda, \eta, \epsilon$  
Initialize: $\theta_0, m_0, v_0 \leftarrow 0$  
while not converged:

1. $g_t \leftarrow \nabla_\theta f(\theta_{t-1})$

2. Update first and second moment estimates:  

   $$m_t \leftarrow \beta_1 m_{t-1} + (1 - \beta_1) g_t \quad ; \quad v_t \leftarrow \beta_2 v_{t-1} + (1 - \beta_2) g_t^2$$

3. Bias corrections:  

   $$\hat{m}_t = \frac{m_t}{1 - \beta_1^t} \quad; \quad \hat{v}_t = \frac{v_t}{1 - \beta_2^t}$$

4. Parameter Update (with decoupled weight decay):  

   $$\theta_t \leftarrow \theta_{t-1}(1 - \eta \lambda)
   - \eta \frac{\hat{m}_t}{\sqrt{\hat{v}_t} + \epsilon}$$

---

### Key Differences

- **Number of Buffers**:  
  - *AdamW* uses two buffers $m, v$ to track first and second moments.  
  - *Lion* uses one main momentum buffer $m$ plus an EMA of gradients $m_t$, but it does **not** scale by $\sqrt{v}$.

- **Normalization**:  
  - *AdamW* divides by $\sqrt{v_t} + \epsilon$.  
  - *Lion* uses the **sign** of the momentum $\mathrm{sign}(c_t)$for its update magnitude (simply $\pm 1$ or $0$).

- **Weight Decay**:  
  - *AdamW* usually uses decoupled weight decay as $\theta \leftarrow \theta (1-\eta\lambda)$.  
  - *Lion* can include a similar decoupled term $\lambda,\theta$ directly in the update step.

Overall, **Lion** often yields sparser updates and requires only a single main momentum buffer, whereas **AdamW** provides an adaptive learning rate through its second moment. 

## 2. Why Did We Implement This?

Lion can sometimes outperform AdamW in terms of:

1. **Convergence Speed**  
   On certain large-scale tasks (e.g., language modeling, image classification), Lion can reach competitive or better validation metrics in fewer iterations.

2. **Memory Savings**  
   Only one momentum buffer is required (instead of two with Adam/AdamW).

3. **Potential for Sharper Minima**  
   The sign-based update can implicitly add a form of regularization that helps generalize well.

Thus, Lion is a relatively new but simple optimizer that can yield speed and accuracy benefits on certain tasks.

### Note
For now LION optimizer is implemented only with cpu kernel, so gpu training is not available with LION optimizer yet.

### CODE REVIEW

File: include/nntile/tensor/lion_step.hh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ {
+ 
+ template<typename T>
+ void lion_step_async(Index num_iter, Scalar beta_1, Scalar beta_2, Scalar eps, Scalar lr, Scalar weight_decay,
+     const Tensor<T> &grad, const Tensor<T> &first_moment, const Tensor<T> &second_moment,
+                    const Tensor<T> &p);
+ 
+ template<typename T>
+ void lion_step(Index num_iter, Scalar beta_1, Scalar beta_2, Scalar eps, Scalar lr, Scalar weight_decay,
+     const Tensor<T> &grad, const Tensor<T> &first_moment, const Tensor<T> &second_moment,
```
Comment (Muxas) [inline include/nntile/tensor/lion_step.hh:29]: remove the second moment

File: include/nntile/tile/lion_step.hh [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+ #include <nntile/tile/tile.hh>
+ 
+ namespace nntile::tile
+ {
+ 
+ // Asynchronous tile-wise Lion step operation
+ template<typename T>
+ void lion_step_async(Index num_iter, Scalar beta_1, Scalar beta_2, Scalar eps, Scalar lr, Scalar weight_decay,
+                      const Tile<T> &grad, const Tile<T> &first_moment, const Tile<T> &second_moment,
```
Comment (Muxas) [inline include/nntile/tile/lion_step.hh:25]: remove the second moment

File: include/nntile/tile/lion_step.hh [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                      const Tile<T> &p);
+ 
+ // Blocking version of tile-wise Lion step operation
+ template<typename T>
+ void lion_step(Index num_iter, Scalar beta_1, Scalar beta_2, Scalar eps, Scalar lr, Scalar weight_decay,
+                const Tile<T> &grad, const Tile<T> &first_moment, const Tile<T> &second_moment,
```
Comment (Muxas) [inline include/nntile/tile/lion_step.hh:31]: remove the second moment

File: src/kernel/lion_step/cpu.cc [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+ #include <cmath>
+ #include "nntile/kernel/cpu.hh"
+ 
+ namespace nntile::kernel::lion_step
+ {
+ 
+ template<typename T>
+ void cpu(Index num_iter, Index num_elems, Scalar beta_1_, Scalar beta_2_,
+         Scalar eps_, Scalar lr_, Scalar weight_decay_, const T *grad,
+         T *first_moment, T *second_moment, T *p)
```
Comment (Muxas) [inline src/kernel/lion_step/cpu.cc:25]: remove the second moment and eps

File: src/kernel/lion_step/cpu.cc [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  * @param[in] num_iter: current iteration number
+  * @param[in] num_elems: Number of elements in buffers
+  * @param[in] beta_1_: parameter for moving average of first moments
+  * @param[in] beta_2_: parameter for moving average of second moments
+  * @param[in] eps_: small scalar to avoid division by zero
+  * @param[in] lr_: learning rate
+  * @param[in] weight_decay_: coefficient for l2 regularizer
+  * @param[in] grad_: Input buffer stored gradient, can be updated if weight_decay > 0
+  * @param[inout] first_moment_: Input buffer stored first moments
+  * @param[inout] second_moment_: Input buffer stored square root of second moments for stability
```
Comment (Muxas) [inline src/kernel/lion_step/cpu.cc:39]: rm the second moment

File: src/kernel/lion_step/cpu.cc [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+         T *first_moment, T *second_moment, T *p)
+     noexcept
+ //! Fused Lion step on buffers
+ /*!
+  *
+  * @param[in] num_iter: current iteration number
+  * @param[in] num_elems: Number of elements in buffers
+  * @param[in] beta_1_: parameter for moving average of first moments
+  * @param[in] beta_2_: parameter for moving average of second moments
+  * @param[in] eps_: small scalar to avoid division by zero
```
Comment (Muxas) [inline src/kernel/lion_step/cpu.cc:34]: rm eps parameter

**Review by Muxas — CHANGES_REQUESTED** [review]
I checked half of the files. You need to remove the second moment and eps parameter. Also, every file ends up with an empty line without end of line symbol. This can be resolved by changing settings of VSCode on the bottom CRLF -> LF. You can also install `pre-commit` through pip to help you automatically check linting of the code.

File: include/nntile/kernel/lion_step/cuda.hh [hunk 7]
```diff
@@ -0,0 +1,22 @@
+#pragma once
```
Comment (Muxas) [inline include/nntile/kernel/lion_step/cuda.hh:1]: Add legal header

File: include/nntile/starpu/lion_step.hh [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+ #include <nntile/starpu/config.hh>
+ 
+ namespace nntile::starpu::lion_step
+ {
+ 
+ //! Structure for arguments
+ struct args_t
+ {
+     Index num_iter;
+     // Index num_elems;
```
Comment (Muxas) [inline include/nntile/starpu/lion_step.hh:27]: Why are these values commented out? They are needed by kernel implementation.

File: src/starpu/lion_step.cc [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+             first_moments, second_moments, p);
+ #endif // STARPU_SIMGRID
+ }
+ 
+ Codelet codelet_fp32, codelet_fp64, codelet_fp32_fast_tf32, codelet_bf16,
+         codelet_fp32_fast_fp16, codelet_fp32_fast_bf16;
+
+void init()
+{
+    codelet_fp32.init("nntile_adam_step_fp32",
```
Comment (Muxas) [inline src/starpu/lion_step.cc:51]: change name of performance model here to lion_step

File: wrappers/python/examples/gpt2_lmhead_training.py [hunk 10]
```diff
@@ -121,6 +121,9 @@
             optimizer = SGD(model_torch.parameters(), args.lr)
         elif args.optimizer == "adamw":
             optimizer = AdamW(model_torch.parameters(), args.lr)
+        # there is no torch implementation for LION, however, I don't see how this optimizer is used
+        elif args.optimizer == "lion":
```
Comment (Muxas) [inline wrappers/python/examples/gpt2_lmhead_training.py:125]: print a statement about no LION implementation in Torch and using ADAMW instead

File: wrappers/python/examples/roberta_training.py [hunk 11]
```diff
@@ -122,6 +122,9 @@
             optimizer = SGD(model_torch.parameters(), args.lr)
         elif args.optimizer == "adamw":
             optimizer = AdamW(model_torch.parameters(), args.lr)
+        elif args.optimizer == "lion":
+             # there is no torch implementation for LION, however, I don't see how this optimizer is used
```
Comment (Muxas) [inline wrappers/python/examples/roberta_training.py:126]: Print this statement to make user aware

File: wrappers/python/nntile/functions.py [hunk 12]
```diff
@@ -1860,6 +1860,105 @@ def fused_adamw_step(
     else:
         raise TypeError
 
+def fused_lion_step(
+    p: Tensor,
+    grad: Tensor,
+    first_moment: Tensor,
+    second_moment: Tensor,
```
Comment (Muxas) [inline wrappers/python/nntile/functions.py:1867]: remove second moment

File: wrappers/python/nntile/model/gpt2.py [hunk 13]
```diff
@@ -320,7 +321,11 @@ def __init__(
         self.num_hidden_layers = num_hidden_layers
         self.next_tag = next_tag
         # Fill Base Model with the generated data
+        print('before super: ', self.config)
         super().__init__(activations, layers)
+        print('after super: ', self.config)
```
Comment (Muxas) [inline wrappers/python/nntile/model/gpt2.py:326]: It is better to use gpt2lmhead model, as gpt2 is deprecated. GPT2 is still here, as not all functionality as moved to GPT2LMHeadModel

File: wrappers/python/nntile/optimizer/lion.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+        self.weight_decay = weight_decay
+        self.eps = eps
+
+    def get_next_tag(self):
+        return self.next_tag
+
+    def unregister(self):
+        for i in range(len(self.first_moments)):
+            self.first_moments[i].unregister()
+            self.second_moments[i].unregister()
```
Comment (Muxas) [inline wrappers/python/nntile/optimizer/lion.py:67]: no second moments

File: wrappers/python/nntile/optimizer/lion.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+        dtype=np.float32,
+        start_lr=None,
+        full_lr_iter=None,
+    ):
+        self.params = params
+        self.next_tag = next_tag
+        self.num_iter = 1
+        self.dtype = dtype
+        self.first_moments = []
+        self.second_moments = []
```
Comment (Muxas) [inline wrappers/python/nntile/optimizer/lion.py:42]: no second moments

**Review by Muxas — CHANGES_REQUESTED** [review]
The code still requires additional work. Seems like some commits are still not here

**Comment by Muxas** [comment]
Fix linting at first. You can use `pre-commit` python package for that purpose. Install it into your environment via `pip install pre-commit` and then do a `pre-commit run --files fname.cc` for your files.
